zkvm/src/types.rs

```diff
@@ -340,6 +339,15 @@ impl DataWitness {
             DataWitness::Input(b) => b.encode(buf),
         }
     }
+
+    fn serialized_len(&self) -> usize {
+        match self {
+            DataWitness::Program(instr) => instr.iter().map(|p| p.serialized_length()).sum(),
```

Comment:
There's no way to avoid this recursion and still get the precise serialized length, but I figure it is alright since it shouldn't be possible to pass in an infinitely-recursing programs as input — not sure if there is a better guard against that!

Reply:
It’s totally fine here: prover creates the tree of types and it’s no big deal to traverse it here.

Reply:
If we use full “_length” elsewhere, why “_len” here? My mild personal preference is to use “_len” to match “len()” as used in collections.